Add support for Digital Credentials on iOS

gp-iaik opened the pull request:
Closes https://github.com/a-sit-plus/valera/issues/163
Requires https://github.com/a-sit-plus/vck/pull/538
Closes https://github.com/a-sit-plus/valera/issues/432

gp-iaik committed:
Add entitlement to app

gp-iaik committed:
First implementation of registering and launching app via DC API on iOS

gp-iaik committed:
WIP for debugging

gp-iaik committed:
Fix loading compose-resources in extension

gp-iaik committed:
Add simple UI to show pre-request

gp-iaik committed:
Add app group/keychain entitlements

gp-iaik committed:
Update project file

gp-iaik committed:
WIP: Update UI for DC API

gp-iaik committed:
Store app data in group for DC API

gp-iaik committed:
WIP: DC API iOS

gp-iaik committed:
Add support for parsing and displaying iOS DC API requests

gp-iaik committed:
Update non-local presentation code to multipaz 0.96

gp-iaik committed:
Add cross-platform support for ISO 18013-7 flow

gp-iaik committed:
Remove unused Swift code

gp-iaik committed:
Add TODOs

gp-iaik committed:
Adapt serialization of IsoMdoc response as iOS does some magic on its own

gp-iaik committed:
Add Xcode changes that just appeared

gp-iaik committed:
Update changelog

gp-iaik committed:
Fix non-macOS build

gp-iaik committed:
Add jvmToolchain and upate Testballoon

gp-iaik committed:
Prepare receiving issuing requests via DC API

gp-iaik committed:
Rework intent navigation

gp-iaik committed:
Use custom tabs, bug fixes

gp-iaik committed:
Remove IntentActivity

gp-iaik committed:
Refactor

gp-iaik committed:
WIP: Refactor intent handling

gp-iaik committed:
Optimise imports

gp-iaik commented:
Requires update of multipaz to latest release

nodh commented on cinterop/DigitalCredentials/DigitalCredentials.swift:
```diff
@@ -0,0 +1,46 @@
+        return true
+    }
+
+    /*@objc public func removeDocuments(_ documentIds: [String]) -> Boolean {
```
still needed?

nodh commented on iosApp/iosApp/ContentView.swift:
```diff
@@ -5,6 +5,9 @@ import shared
     
     func makeUIViewController(context: Context) -> UIViewController {
+        Napier.shared.base(antilog:OSLogNapierAntilog())
+        Napier.shared.log(priority: LogLevel.debug, tag: "123", throwable: nil, message: "HIER!!!!!!!!!!!!!!!!!!!!!!!!!33")
```
still needed?

nodh commented on shared/src/androidMain/kotlin/matcher/Request.cpp:
```diff
@@ -28,6 +28,14 @@ std::string base64UrlDecode(const std::string& data) {
+std::unique_ptr<MdocRequest> MdocRequest::parseSomething(const std::string& protocolName, cJSON* dataJson) {
+    auto dataElements = std::vector<MdocRequestDataElement>();
+    dataElements.push_back(MdocRequestDataElement("namespaceName", "dataElementName", false));
+    dataElements.push_back(MdocRequestDataElement("namespaceName2", "dataElementName2", false));
```
is this needed?

nodh commented on shared/src/commonMain/kotlin/at/asitplus/wallet/app/common/domain/communications/di/CommunicationsModule.kt:
```diff
@@ -11,12 +11,13 @@ import org.koin.core.module.dsl.singleOf
     singleOf(::SigningService)
-    singleOf(::DCAPIExportService)
+    single { DCAPIExportService(get(), shouldListen) }
+    //singleOf(::DCAPIExportService)
```
still needed?

nodh commented on shared/src/iosMain/kotlin/OsLogAntilog.kt:
still needed?

gp-iaik committed:
WIP: Add DC API issuing code

gp-iaik committed:
Add missing methods

gp-iaik committed:
Fix returning ISO Mdoc response with new DC API lib

gp-iaik committed:
Rename DC API verify method

gp-iaik committed:
Parse credential offer

gp-iaik committed:
Handle issuance in UI

gp-iaik committed:
Restore issuing registration

gp-iaik committed:
Add stub methods on iOS

gp-iaik committed:
Parse received issuer metadata

gp-iaik committed:
Add issuance matcher

gp-iaik committed:
Use single cJSON source for both matcher

gp-iaik committed:
Return response to caller

gp-iaik committed:
Use supplied authorizationServerMetadata

gp-iaik committed:
Update changelog

gp-iaik committed:
Cleanup result handling

gp-iaik committed:
Fix back button behaviour

gp-iaik committed:
Remove unused dependency

gp-iaik committed:
Update matcher

gp-iaik committed:
Update multipaz usage

gp-iaik committed:
Update dependencies

gp-iaik committed:
Use hardcoded matcher

gp-iaik committed:
Delete newer issuance matcher

gp-iaik committed:
Add README on how to build matcher

gp-iaik committed:
Add failed navigation warning to log

gp-iaik committed:
Simplify exception behaviour

gp-iaik committed:
Change finishApp behaviour for iOS

gp-iaik committed:
Simplify DC API issuing exception handling

gp-iaik committed:
Rename creation to issuing

gp-iaik committed:
Cleanup

gp-iaik committed:
Update vck

gp-iaik committed:
Fix iOS build

gp-iaik committed:
Update after rebase

gp-iaik committed:
Add error handling for init

gp-iaik committed:
Require checking for permissions for transfer methods set by user

gp-iaik commented:
TODO: fix proximity

gp-iaik committed:
Merge remote-tracking branch 'origin/development' into feature/ios_digital_credentials

gp-iaik committed:
Update dependencies

gp-iaik committed:
Cleanup

gp-iaik committed:
Support complex DCQL queries

gp-iaik committed:
Set filterByIds

gp-iaik committed:
Use invocation aware back handler for all cross-app authentication flows

gp-iaik committed:
Update multipaz to 0.98.0

gp-iaik committed:
Add success screen for DC API issuing

gp-iaik committed:
Fix test execution

gp-iaik committed:
Check pre-request consistency

gp-iaik committed:
Fix some iOS build issues

gp-iaik committed:
Some more iOS build issues

gp-iaik committed:
Merge remote-tracking branch 'origin/development' into feature/ios_digital_credentials

gp-iaik committed:
Cleanup

gp-iaik committed:
Update DCAPI request handling for credentialIds

gp-iaik committed:
Merge branch 'feature/ios_digital_credentials' of github.com:a-sit-plus/compose-wallet-app into feature/ios_digital_credentials

nodh commented on androidApp/src/androidMain/kotlin/at/asitplus/wallet/app/android/AbstractWalletActivity.kt:
```diff
@@ -40,6 +42,7 @@ abstract class AbstractWalletActivity : AppCompatActivity() {
 
     @OptIn(ExperimentalDigitalCredentialApi::class)
     fun sendCredentialResponseToDCAPIInvoker(resultStr: String, success: Boolean) {
+        // TODO use credentials API error types
```
Tracked in https://github.com/a-sit-plus/valera/issues/440

nodh commented on shared/src/iosMain/kotlin/main.ios.kt:
```diff
@@ -252,23 +360,68 @@ class IosPlatformAdapter(
+
+    override fun prepareDCAPICredentialResponse(response: String, success: Boolean) {
+        Napier.w("Got error response: $response")
+        //TODO is there a way to convey error responses via ISO18013-7?
```
Tracked in https://github.com/a-sit-plus/valera/issues/440

nodh commented on shared/src/iosMain/kotlin/main.ios.kt:
```diff
@@ -252,23 +360,68 @@ class IosPlatformAdapter(
-        return KmmResult.failure(Throwable("Using Swift platform adapter"))
+    override fun getCurrentDCAPIVerificationData(): KmmResult<DCAPIWalletRequest> {
+        Napier.d("getCurrentDCAPIData called")
+        // TODO update code so that getCurrentDCAPIData is invoked, i.e. open the right views
```
Relevant?

gp-iaik replied:
Not anymore

nodh commented on shared/src/iosMain/kotlin/main.ios.kt:
```diff
@@ -252,23 +360,68 @@ class IosPlatformAdapter(
+                require(parsedRequestSummary.isConsistentWith(isoMdocRequest)) {
+                    "Parsed ISO18013 mobile document pre-request is inconsistent with rawRequest"
+                }
+                //TODO check added nullability of DCAPIWalletRequest in Android code and terminal sp
```
Relevant?

gp-iaik replied:
No, already checked

nodh commented on shared/src/iosMain/kotlin/main.ios.kt:
```diff
@@ -80,14 +73,34 @@ actual fun getColorScheme(): ColorScheme {
+fun getIosIntentState(): IntentState = iosIntentState
+
+object MdocSessionManager {
+    // TODO check if correct credentials are shown without credentialId set (and check behaviour on Android, should only show the one credential selected by the user)
```
Relevant?

nodh replied:
https://github.com/a-sit-plus/valera/issues/444

nodh commented on shared/src/commonMain/kotlin/ui/viewmodels/intents/DCAPIIssuingIntentViewModel.kt:
```diff
@@ -0,0 +1,44 @@
+        val creationOptions =
+            vckJsonSerializer.decodeFromString<DigitalCredentialCreationOptions>(requestJson)
+        // TODO specification does not yet define what to do with multiple requests
+        // TODO parse and check origin once its sent by the web platform ("… will send the credential offer along with the Origin of the Issuer to the End-User's chosen Wallet.")
```
Relevant? Do we need an tracking issue?

nodh replied:
https://github.com/a-sit-plus/valera/issues/443

nodh commented on shared/src/commonMain/kotlin/at/asitplus/wallet/app/common/iso/transfer/TransferManager.kt:
```diff
@@ -143,16 +144,28 @@ class TransferManager(
+            serial = ASN1Integer(1L),
+            validFrom = LocalDate.parse("2025-06-26").atTime(10, 0).toInstant(TimeZone.UTC),
+            validUntil = LocalDate.parse("2027-06-26").atStartOfDayIn(TimeZone.UTC),
+            dnsName = "a-sit.at",
```
```suggestion
            dnsName = "wallet.a-sit.at",
```

nodh commented on shared/src/commonMain/kotlin/at/asitplus/wallet/app/common/domain/communications/di/CommunicationsModule.kt:
```diff
@@ -11,12 +11,12 @@ import org.koin.core.module.dsl.singleOf
 import org.koin.dsl.module
 
-fun communicationsModule() = module {
+fun communicationsModule(shouldListen: Boolean) = module {
```
should listen to what?

gp-iaik replied:
Debugging remnant

nodh commented on shared/src/androidMain/kotlin/matcher_issuance_hardcoded/README.md:
Should we include a pointer to this readme in our `DEVELOPMENT.md`? When do we need to perform these steps?

gp-iaik replied:
Only when touching the code in the matcher

nodh commented on iosApp/IdentityDocumentProviderExtension/DocumentProviderExtension.swift:
```diff
@@ -0,0 +1,193 @@
+                Task {
+                    do {
+                        try await requestContext.sendResponse { rawRequest in
+                            // TODO show pre-request inside our UI so that we can use user-friendly names for the requested attributes and
```
Todo relevant?

gp-iaik replied:
Yes, should be tracked in an issue

nodh replied:
https://github.com/a-sit-plus/valera/issues/442

nodh reviewed (changes requested):
Great! Can I test it against https://digital-credentials.dev/?

Pending publishing of VC-K in https://github.com/a-sit-plus/vck/actions/runs/23642711490

gp-iaik committed:
Update dnsName for proximity presentation

Co-authored-by: Christian Kollmann <christian.kollmann@a-sit.at>

gp-iaik committed:
Cleanup

gp-iaik commented:
> Great! Can I test it against https://digital-credentials.dev/?

Good idea, because I didn't yet test it after merging in the iOS development changes on my slow Mac. You need a verifier that supports ISO 18013-7 Annex C though. https://tools.mattrlabs.com/verify-credentials supports it I think.

gp-iaik committed:
Update DEVELOPMENT.md

gp-iaik committed:
Merge branch 'feature/ios_digital_credentials' of github.com:a-sit-plus/compose-wallet-app into feature/ios_digital_credentials

gp-iaik committed:
Resolve call ambiguity

gp-iaik committed:
Add potential fix for iOS runner

gp-iaik committed:
Add another potential fix for iOS runner

gp-iaik committed:
Possibly fix iOS runner

gp-iaik committed:
Remove TODO

gp-iaik committed:
Merge remote-tracking branch 'origin/development' into feature/ios_digital_credentials

nodh commented:
@gp-iaik `Task :shared:linkDebugTestIosSimulatorArm64 FAILED` in https://github.com/a-sit-plus/valera/actions/runs/23798931309/job/69353598670?pr=416

nodh commented:
@codex review

gp-iaik commented:
> @gp-iaik `Task :shared:linkDebugTestIosSimulatorArm64 FAILED` in https://github.com/a-sit-plus/valera/actions/runs/23798931309/job/69353598670?pr=416

@n0900  is working on it

gp-iaik committed:
Add data store migration, read app group identifier from config

gp-iaik committed:
Merge branch 'feature/ios_digital_credentials' of github.com:a-sit-plus/compose-wallet-app into feature/ios_digital_credentials

gp-iaik committed:
Merge remote-tracking branch 'origin/development' into feature/ios_digital_credentials

gp-iaik committed:
Test with XCode 26

gp-iaik committed:
Update linker

gp-iaik committed:
Use feature/bumpXcode

gp-iaik committed:
Remove hardcoded Xcode paths

nodh commented on interop/build.gradle.kts:
```diff
@@ -0,0 +1,76 @@
+kotlin {
+    val isMacHost = System.getProperty("os.name").lowercase().contains("mac")
+    val developerDir = if (isMacHost) {
+        project.providers.environmentVariable("DEVELOPER_DIR")
```
What value would I need to set here?

gp-iaik replied:
This can be set to the Xcode directory, as shown in the examples at https://manpagez.com/man/1/xcode-select/

nodh commented:
@codex review

gp-iaik commented:
Still not building because linking of the Multipaz interop bridge fails :(

nodh committed:
Update test-ios.yml

nodh committed:
Merge branch 'development' into feature/ios_digital_credentials

gp-iaik committed:
Update conventions

nodh committed:
Build: Remove compose repo

nodh committed:
Build: Try to fix iOS build

gp-iaik committed:
Downgrade testballon to vck version

Possible fix for unlinked class symbols in CI runner

gp-iaik committed:
Add possible fix for missing libswift_Concurrency dynamic library

nodh committed:
Build: Try to load concurrency lib for simulator run

nodh committed:
Build: Try to load concurrency lib for simulator run

gp-iaik committed:
Create TestDataStore for iOS

gp-iaik committed:
Merge branch 'feature/ios_digital_credentials' of github.com:a-sit-plus/compose-wallet-app into feature/ios_digital_credentials

gp-iaik committed:
Add fix for TestDataStore

gp-iaik committed:
Add debug logging to tests

gp-iaik committed:
Add network diagnostic test

gp-iaik committed:
Possibly fix duplicate libswift_Concurrency

gp-iaik committed:
Fix deprecation warning, reduce app size

gp-iaik committed:
Import used icons only

gp-iaik committed:
Fix material3 version

nodh commented:
@codex review and analyze the build error on iOS too

gp-iaik committed:
Revert "Possibly fix duplicate libswift_Concurrency"

This reverts commit bff96d266b9f6a9c3354e37e321bf5aa1c4eb7d7.

JesusMcCloud committed:
try with latest code and iOS 18 (#447)

* Xcode 26

* remove xcode <26 workarounds

* no standalone simulator

* Raise the iOS target version to 18.6

JesusMcCloud committed:
Update at.asitplus.gradle.conventions

nodh commented on gradle/libs.versions.toml:
```diff
@@ -3,21 +3,24 @@ vck="5.12.0-SNAPSHOT"
 koin = "4.1.1"
-testballoon = "0.7.1-K2.3.0"
+# TestBallon has to be the same version as in vck, otherwise tests in composite builds may fail
+testballoon = "0.8.2-K2.3.0"
```
@JesusMcCloud isnt that covered by the conventions?

JesusMcCloud replied:
latest and greatest conventiosn sets these:

```
testballoon=0.8.4-K2.3.0
testballoonAddons=0.8.0
```

JesusMcCloud replied:
but there are certain versions that require strict alignment, because composite builds with differing version catalogs override versions and this can result in a mismatch. omitting the versions altogether works as long as they remain aligned with the kotlin version

nodh commented on shared/src/commonMain/kotlin/ui/viewmodels/LoadCredentialViewModel.kt:
```diff
@@ -67,6 +97,30 @@ class LoadCredentialViewModel(
+            onClickLogo = onClickLogo,
+            onClickSettings = onClickSettings,
+            credentialIdentifiers = walletMain.scope.async {
+                val issuerMetadata = requireNotNull(offer.credentialIssuerMetadata) {
+                    "Missing credential issuer metadata for DC API request"
+                }
```
@gp-iaik relevant?

gp-iaik replied:
DC API requires credential issuer metadata as per pre-spec

nodh commented on shared/src/iosTest/kotlin/at/asitplus/wallet/app/IosNetworkingDiagnosticsTest.kt:
Can be removed now

nodh commented:
@codex review

gp-iaik committed:
Revert "Add network diagnostic test"

This reverts commit 7bf64768a886df6038fde971932fe58d5cee77a9.

gp-iaik committed:
Merge branch 'feature/ios_digital_credentials' of github.com:a-sit-plus/compose-wallet-app into feature/ios_digital_credentials

gp-iaik commented on iosApp/iosApp.xcodeproj/project.pbxproj:
```diff
@@ -393,6 +554,104 @@
+				INFOPLIST_FILE = IdentityDocumentProviderExtension/Info.plist;
+				INFOPLIST_KEY_CFBundleDisplayName = IdentityDocumentProviderExtension;
+				INFOPLIST_KEY_NSHumanReadableCopyright = "Copyright © 2025 orgName. All rights reserved.";
+				IPHONEOS_DEPLOYMENT_TARGET = 26.0;
```
🤷

gp-iaik commented:
> ### 💡 Codex Review
> 
> https://github.com/a-sit-plus/valera/blob/dd1aae7e8d97bc4ba29f8e3a886224da4c412eb3/shared/src/commonMain/kotlin/ui/navigation/WalletNavigation.kt#L891-L893
> 
> **![P1 Badge](https://camo.githubusercontent.com/c595229c0ecb6ee85b9c7804144d495f131a495ec87091fea2b262d954c9a92d/68747470733a2f2f696d672e736869656c64732e696f2f62616467652f50312d6f72616e67653f7374796c653d666c6174) Navigate to returned route after provisioning resume**
> 
> `ProvisioningIntentViewModel` now returns an optional route on success (including `AddCredentialDcApiSuccessRoute` when handling `CREATE_CREDENTIAL` flows), but this callback ignores the route and always calls `navigateBack()`. In issuer flows that require browser auth (`resumeWithAuthCode`), the app will return to the previous screen instead of the DCAPI success screen, so the wallet never sends the final create-credential response back to the invoker.
> ℹ️ About Codex in GitHub

Should be irrelevant for DC API.

nodh committed:
Merge branch 'development' into feature/ios_digital_credentials

gp-iaik committed:
Restore document registration on iOS

gp-iaik committed:
Merge branch 'feature/ios_digital_credentials' of github.com:a-sit-plus/compose-wallet-app into feature/ios_digital_credentials

nodh committed:
Build: Fix building on iOS

nodh committed:
iOS: Fix DCAPI parsing

gp-iaik committed:
Remove workaround which is not required anymore

gp-iaik committed:
Update credentials library

nodh committed:
DCAPI: Be sure to normalize origin

gp-iaik committed:
Separate acitivity and Koin objects for sharing/main app, lots of UI fixes (#452)

* Fix Compose guideline violations in navigation and initialization

  - Replace bare `var pendingRoute` with `mutableStateOf` so the pending
    prerequisites route survives recomposition
  - Wrap InitializationView navigation calls in LaunchedEffect to prevent
    repeated side-effects during recomposition
  - Replace unscoped CoroutineScope(Dispatchers.Main) with
    rememberCoroutineScope() so navigation coroutines are cancelled when
    the composable leaves the composition
  - Replace runBlocking in LoadCredential* composables with
    LaunchedEffect + mutableStateOf, eliminating main-thread blocking
    during async ViewModel initialisation

* WIP: Try to fix multiple app instances

* Use flow-local bindings for Koin objects to avoid state issues between normal and sharing app instances

* iOS

* Fix nav issues

* Remove legacy entrypoint

* Bind coroutine scopes to current session

* Adapt NdefService for new lifecycle management and cleanup startup code

* Fix typo in strings.xml

* Add error message for DC API registration, fix entitlements

* Add finishApp on iOS

* Add finishApp on iOS

* WIP: Fix iOS DC API navigation

* WIP: Logging changes

TODO: fix logging for non-iOS builds

* iOS navigation changes

* WIP: iOS build changes

* DCAPI: Add some basic tests

* DCAPI: Test device namespace payload stability

* DCAPI: Use ephemeral key for iOS tests

* Set platform-specific antilog adapter

* WIP: Potential fix for 2nd request

* Go to the app's system settings page on permission denial

* Unify navigation between main and sharing flows

* Fix test build

* Fix multiple navigation issues in tab view, remove deprecated BackHandler

* Move test to commonMain

* Do not cancel the flow when the user does not unlock key material

* Show requested attributes for Annex C requests

* Back/Close handling in sharing flow

* Use same back/cancel handling for URL flow, exclude sharing activity from recents

* More navigation fixes

* Move signing flow to SharingActivity

* Move URL-based issuing to transient flow

* Show issued credential on success screen

* Fix potential issues

* Handle incoming links on iOS in transient flow too

* Show more informative heading

* Audit fixes

* Add potential fix for ClassCastException crash

* Fix showing error messages to the user

* Fix more instances of this error pattern

* Rename returnToHome to something more apt

* Update dependencies

* Fix koin initialisation for CoreResolverV2

* Fix tests

* Add spacing

* Fix iOS DC API navigation

* Add close button

* Fix Compose guideline violations in navigation and initialization

  - Replace bare `var pendingRoute` with `mutableStateOf` so the pending
    prerequisites route survives recomposition
  - Wrap InitializationView navigation calls in LaunchedEffect to prevent
    repeated side-effects during recomposition
  - Replace unscoped CoroutineScope(Dispatchers.Main) with
    rememberCoroutineScope() so navigation coroutines are cancelled when
    the composable leaves the composition
  - Replace runBlocking in LoadCredential* composables with
    LaunchedEffect + mutableStateOf, eliminating main-thread blocking
    during async ViewModel initialisation

* WIP: Try to fix multiple app instances

* Use flow-local bindings for Koin objects to avoid state issues between normal and sharing app instances

* iOS

* Fix nav issues

* Remove legacy entrypoint

* Bind coroutine scopes to current session

* Adapt NdefService for new lifecycle management and cleanup startup code

* Fix typo in strings.xml

* Add error message for DC API registration, fix entitlements

* Add finishApp on iOS

* Add finishApp on iOS

* WIP: Fix iOS DC API navigation

* WIP: Logging changes

TODO: fix logging for non-iOS builds

* iOS navigation changes

* WIP: iOS build changes

* Set platform-specific antilog adapter

* WIP: Potential fix for 2nd request

* Go to the app's system settings page on permission denial

* Unify navigation between main and sharing flows

* Fix test build

* Fix multiple navigation issues in tab view, remove deprecated BackHandler

* Move test to commonMain

* Do not cancel the flow when the user does not unlock key material

* Show requested attributes for Annex C requests

* Back/Close handling in sharing flow

* Use same back/cancel handling for URL flow, exclude sharing activity from recents

* More navigation fixes

* Move signing flow to SharingActivity

* Move URL-based issuing to transient flow

* Show issued credential on success screen

* Fix potential issues

* Handle incoming links on iOS in transient flow too

* Show more informative heading

* Audit fixes

* Add potential fix for ClassCastException crash

* Fix showing error messages to the user

* Fix more instances of this error pattern

* Rename returnToHome to something more apt

* Update dependencies

* Fix koin initialisation for CoreResolverV2

* Fix tests

* Add spacing

* Fix iOS DC API navigation

* Fix after rebase

* Add more tests for pre-request consitency

* Do not show summary screen after pre-request screen

* Fix double tap on settings icon on share flow

* Fix warning

* Cleanup loading of home screen

* Fix merge duplicate

* Fix camera permission handling and unify QR code scanner interfaces

* Show a snackbar instead of a full-blown dialog for credential refresh

* WIP: Local presentment state handling changes

* Fix race conditions when enabling/disabling Bluetooth

* Retry on lost NFC tag

* Fix engagement -> NFC data transfer handover, multiple state/UX fixes

* Adapt timeout to match specification

* Fix NFC data transfer handover

* More local presentation fixes and usability improvements

* Fixes for local presentment

* Fix NFC engagement on iOS

* Fix QR -> NFC data transfer flow

* Suppress new tag popup on result screen

* Fix continuation exception on error

* Add comments

* Cleanup

* Add error view for Android-native coroutine scopes

* Handle no matching credential dismiss case

* iOS: Fix local device build

* Provisioning: Fix code and challenge sent to server on auth code flows

* Remove browser from history to avoid jumping back to it

---------

Co-authored-by: Christian Kollmann <christian.kollmann@a-sit.at>

gp-iaik committed:
Merge remote-tracking branch 'origin/development' into feature/ios_digital_credentials

gp-iaik committed:
Fix CredentialAliditySerice

gp-iaik committed:
Merge remote-tracking branch 'origin/development' into feature/ios_digital_credentials

gp-iaik committed:
catchingUnwrapped

gp-iaik committed:
Adapt tests

gp-iaik committed:
Adapt tests

gp-iaik committed:
WIP: Adapt for iOS signum change

gp-iaik committed:
Merge remote-tracking branch 'origin/feature/ios_digital_credentials' into feature/ios_digital_credentials

# Conflicts:
#	shared/src/commonTest/kotlin/at/asitplus/wallet/app/InstrumentedTestsSuite.kt

gp-iaik committed:
Fix IdentityDocumentProviderExtension bundle ID to derive from APP_BUNDLE_IDENTIFIER

Hardcoded bundle ID broke embedding validation whenever APP_BUNDLE_IDENTIFIER was overridden in local/release signing configs.

gp-iaik commented on androidApp/src/androidMain/AndroidManifest.xml:
```diff
@@ -43,151 +44,165 @@
+
+        <activity
+            android:name=".TransientFlowActivity"
+            android:configChanges="orientation|screenSize|screenLayout|keyboardHidden|mnc|colorMode|density|fontScale|fontWeightAdjustment|keyboard|layoutDirection|locale|mcc|navigation|smallestScreenSize|touchscreen|uiMode"
```
The general claim is correct but does not apply here.

  The classic singleTask + startActivityForResult problem is real and well-documented: if app A calls startActivityForResult(intent) targeting a singleTask activity in a different task, Android immediately delivers RESULT_CANCELED to A before the target even runs, because cross-task startActivityForResult
  is not supported.

  But Credential Manager does not use startActivityForResult.

  Credential Manager (and RegistryManager for the DC API registry path) invokes provider activities via PendingIntent.send() with a result callback. That goes through ActivityTaskManager via Binder — a system-level cross-process/cross-task result channel, completely separate from the app-level
  startActivityForResult/onActivityResult mechanism. The singleTask restriction only applies to the latter. The framework's Binder-backed result channel is unaffected by task boundaries, so setResult(RESULT_OK, resultData) in sendCredentialResponseToDCAPIInvoker/sendCredentialCreationResponseToDCAPIInvoker
   does reach Credential Manager correctly.

  So the empirical observation is right, and for the right reason — not luck.

  There is one genuine interaction with singleTask worth watching in your code: because singleTask reuses a running instance (calling onNewIntent instead of creating a fresh one), a second credential request arriving while one is already in progress will replace intentState.dcapiInvocationData via
  populateLink (TransientFlowActivity.kt:55–98). The current code handles this correctly by resetting state, but it does mean any in-flight credential presentation is silently abandoned if the activity is reused. That's probably acceptable given excludeFromRecents="true" and the transient-flow purpose, but
   worth being deliberate about.

  The singleTask + separate taskAffinity setup here is intentional and safe for the DC API paths.

gp-iaik committed:
Throw AppReset on iOS if old key is present

gp-iaik committed:
Fix test which failed because snackbar was obstructing buttons

nodh commented on iosApp/IdentityDocumentProviderExtension/IdentityDocumentProviderExtension.entitlements:
```diff
@@ -0,0 +1,10 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
+<plist version="1.0">
+<dict>
+	<key>com.apple.security.application-groups</key>
+	<array>
+		<string>group.Compose.Wallet</string>
+	</array>
```
@gp-iaik is this valid?

gp-iaik committed:
Fix app reset via exception

gp-iaik committed:
Add group identifiers and tag override

gp-iaik committed:
Fix app reset on error #2

gp-iaik committed:
Adapt to unpfusch signum commit

gp-iaik committed:
Adapt to vck DC API changes

gp-iaik committed:
Update conventions version to 20260610

gp-iaik committed:
Merge remote-tracking branch 'origin/development' into feature/ios_digital_credentials

gp-iaik committed:
Adapt old key handling

gp-iaik committed:
Undo keystore changes now that

gp-iaik committed:
Make input fields for URLs show the URL keyboard

gp-iaik committed:
Downgrade navigationevent-compose to fix iOS crash

gp-iaik committed:
Fix capabilities check on issuing via QR, also add them for signing and auth

gp-iaik committed:
Small change

gp-iaik committed:
Merge branch 'feature/ios_digital_credentials' of github.com:a-sit-plus/valera into feature/ios_digital_credentials

gp-iaik committed:
Change default issuer to new instance

gp-iaik committed:
Remove testbaloon from valera and see what happens

gp-iaik committed:
Revert "Remove testbaloon from valera and see what happens"

This reverts commit 3a648f22b4046ab0f5fe586691c1ca1f2d92f82e.

gp-iaik committed:
Update testballoon to Bernds version

gp-iaik committed:
Add potentially missing dependency

gp-iaik committed:
Assign new KS_ALIAS only on iOS to avoid triggering reset on Android

gp-iaik committed:
Update testballoon to vck version

gp-iaik committed:
Merge remote-tracking branch 'origin/development' into feature/ios_digital_credentials

gp-iaik committed:
Add comment

gp-iaik committed:
Improve no credential view

nodh committed:
UI: Add toggle for Wallet Attestation (#467)

gp-iaik committed:
Move attestation off setting to attestation settings

gp-iaik committed:
Add loading message

nodh committed:
Update to VC-K 6.0.0 (#470)

gp-iaik committed:
Fix loading loading message on iOS

gp-iaik committed:
Add potential fix for Android's Bundle-to-JSON conversion

gp-iaik committed:
Adapt logging

gp-iaik committed:
Change translation

gp-iaik committed:
Merge branch 'feature/ios_digital_credentials' of github.com:a-sit-plus/compose-wallet-app into feature/ios_digital_credentials

gp-iaik committed:
Add workaround for Android JSON array issue

nodh committed:
Document correct iOS app signing

nodh committed:
iOS: Open URL on background thread

nodh committed:
Provisioning: Store context per state entry

gp-iaik committed:
iOS session handling improvements
